packages/lib/src/types/RecursiveRecord.d.ts

```diff
@@ -0,0 +1,34 @@
+/**
+ * Create a recursive Record type of any number of key sets.
+ *
+ * The second parameter (`Strict`) determines whether each union of Keys
+ * must be complete in each recursive `Record`
+ * Passing `false` to this parameter will wrap each recursive `Record` in `Partial`,
+ * resulting in a looser Record definition.
+ *
+ * E.g.
+ * ```ts
+ * ColorRecord = RecursiveRecord<
+ *   [Theme, Property, Variant, State, string],
+ *   false
+ * > =
+ * Partial<Record<Theme,
+ *   Partial<Record<Property,
+ *     Partial<Record<Variant,
+ *       Partial<Record<State, string>>
+ *     >>
+ *   >>
+ * >>
+ * ```
+ * */
+export type RecursiveRecord<
+  Keys extends Array<any>,
+  Strict extends boolean = true,
+> = Keys extends [
+  infer Key, // the current union of keys
```

Comment:
Can you ELI5 what's happening on lines 27-30?

Reply:
> If the type `Keys` is an array, then de-structure that type. The first index in the array is an inferred `Key` type, and then a spread the remaining indexes into an inferred `Rest` type.

Reply:
Technically `Keys` is not an array, but a [Variadic Tuple](https://typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types)

Reply:
That was very helpful; thanks for the breakdown!